tests/main/views/test_buyer_domains.py

```diff
@@ -0,0 +1,186 @@
+        assert data['buyerEmailDomains']['domainName'] == new_domain
+
+    @pytest.mark.parametrize(
+        'existing_domain', ['approved.gov.uk', 'already.approved.gov.uk', 'definitely-approved.gov.uk']
+    )
+    def test_higher_level_domain_can_be_created_if_more_specific_domain_already_exists(self, existing_domain):
+        with self.app.app_context():
+            db.session.add(BuyerEmailDomain(domain_name=existing_domain))
+            db.session.commit()
+
+        res = self.client.post(
+            '/buyer-email-domain',
+            data=json.dumps({
+                'buyerEmailDomains': {'domainName': 'gov.uk'},
+                'updated_by': 'example'
+            }),
+            content_type='application/json')
+        data = json.loads(res.get_data(as_text=True))
+
+        assert res.status_code == 201
+        assert data['buyerEmailDomains']['domainName'] == "gov.uk"
+
+    def test_create_buyer_email_domain_forces_to_lower_case(self):
+        res = self.client.post(
+            '/buyer-email-domain',
+            data=json.dumps({
+                'buyerEmailDomains': {'domainName': 'EXAMPLE.org'},
+                'updated_by': 'example'
+            }),
+            content_type='application/json')
+        data = json.loads(res.get_data(as_text=True))
+
+        assert res.status_code == 201
+        assert data['buyerEmailDomains']['domainName'] == "example.org"
+
+    def test_create_buyer_email_domain_fails_if_required_field_is_not_provided(self):
+        res = self.client.post(
+            '/buyer-email-domain',
+            data=json.dumps({
+                'updated_by': 'example'
+            }),
+            content_type='application/json')
+        data = json.loads(res.get_data(as_text=True))
+
+        assert res.status_code == 400
+        assert data['error'] == "Invalid JSON must have 'buyerEmailDomains' keys"
+
+    def test_create_buyer_email_domain_creates_audit_event(self):
+        self.client.post(
+            '/buyer-email-domain',
+            data=json.dumps({
+                'buyerEmailDomains': {'domainName': "example.com"},
+                'updated_by': 'example user'
+            }),
+            content_type='application/json'
+        )
+
+        with self.app.app_context():
+            buyer_email_domain = BuyerEmailDomain.query.filter(
+                BuyerEmailDomain.domain_name == "example.com"
+            ).first()
+
+            audit = AuditEvent.query.filter(
+                AuditEvent.object == buyer_email_domain
+            ).first()
+
+            assert audit.type == "create_buyer_email_domain"
+            assert audit.user == "example user"
+            assert audit.data == {
+                'buyerEmailDomainId': buyer_email_domain.id,
+                'buyerEmailDomainJson': {'domainName': 'example.com'}
+            }
+
+    @pytest.mark.parametrize(
+        'new_domain', ['no-dots', 'at@symbol.org', 'contains space.org', 'bad!char.org', 'tld-too-short.x']
```

Comment:
Would also be nice to have a test case for a plain TLD like `.com`, which shouldn't be allowed.  And maybe we don't allow dots at the start at all?  e.g. should `.gov.uk` fail but `gov.uk` be allowed?

Reply:
Done, as per the comment above